DietPi-Software | K3s

mortenlj opened the pull request:
**Status**: WIP
- [x] Still needs testing
- [x] Needs docs

**Reference**: https://github.com/MichaIng/DietPi/issues/3772

**Commit list/description**:
- DietPi-Software | Added K3s

First shot at this, open to feedback.

I realised that my approach outlined in the issue wouldn't actually work. It is absolutely needed to be able to override the `K3S_EXEC` variable at install time, and in order to do that unattended, I guess it needs to be added to `/boot/dietpi.txt`. The other install-options can be ignored, if you need those you are going to be mucking about anyway, so might as well do it manually.

Also, after it is installed, it needs a configuration, and I opted to allow the user to provide one for the unattended install use case. K3s can run without it, but most people would need to make some changes to it, if they plan to do anything serious. I'm open to taking it out again if you absolutely don't like that pattern. It would naturally be clearly pointed out in the docs.

Next steps after getting some feedback would be to test that it works as intended, and write up some docs.

mortenlj committed:
Add k3s to dietpi-software

Joulinar commented on dietpi/dietpi-software:
```diff
@@ -886,6 +886,13 @@ INDEX_BROWSER_TARGET=$INDEX_BROWSER_TARGET"
 		aSOFTWARE_DOCS[$software_id]='https://dietpi.com/docs/software/cloud/#vaultwarden'
 		aSOFTWARE_DEPS[$software_id]='16 87'
+		#------------------
+		software_id=126
```
Pls use software ID 192

mortenlj replied:
Done

mortenlj committed:
Use 192 as software id

mortenlj committed:
Fix bad indents

mortenlj committed:
Config needs to be in place before install

mortenlj committed:
Be more informative

MichaIng commented on dietpi.txt:
```diff
@@ -260,6 +260,13 @@ SOFTWARE_HOMEASSISTANT_APT_DEPS=
+# The value of this variable is copied directly into the INSTALL_K3S_EXEC environment variable before
+# running the k3s installer.
+# https://rancher.com/docs/k3s/latest/en/installation/install-options/#options-for-installation-with-script
+SOFTWARE_K3S_EXEC=
```
Shall we add a note about the `/boot/dietpi-k3s.yaml` possibility here as well? When users edit `dietpi.txt` already, nice to know files files you can pre-create for installs 🙂.

mortenlj replied:
Yeah, good point. I'll add it.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -886,6 +886,13 @@ INDEX_BROWSER_TARGET=$INDEX_BROWSER_TARGET"
+		#------------------
+		software_id=192
+
+		aSOFTWARE_NAME[$software_id]='k3s'
```
Let's spell it `K3s` is capital K.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -886,6 +886,13 @@ INDEX_BROWSER_TARGET=$INDEX_BROWSER_TARGET"
+
+		aSOFTWARE_NAME[$software_id]='k3s'
+		aSOFTWARE_DESC[$software_id]='The certified Kubernetes distribution built for IoT & Edge computing'
+		aSOFTWARE_CATX[$software_id]=4
```
I'd say it fits better into `'●─ System Stats & Management ' #8` next to Docker, doesn't it?

mortenlj replied:
Category was one of the things I had trouble deciding on. Didn't even realize docker was in that category, but yeah, that makes sense. I'll move it.

mortenlj committed:
Fixes from PR feedback

nonnorm commented on dietpi/dietpi-software:
```diff
@@ -5803,6 +5811,29 @@ Package: wireguard wireguard-dkms wireguard-tools\nPin: release n=bullseye\nPin-
+			fi
+
+			# Install
+			G_EXEC curl -sSfL "$INSTALL_URL_ADDRESS" -o install.sh
```
Why not use `Download_Install` here?
```
Download_Install 'https://get.k3s.io' install.sh
```

MichaIng replied:
Overkill, I'd say. For single files with a known fixed URL address, Download_Install has no advantage.

**EDIT: ... when it's temporary files/installers. For files stored on disk, Download_Install automatically pre-creates the target directory, downloads first and stores the file afterwards (to not overwrite a file when the download may break) and prints an info that the original file is updated etc 🙂.**

MichaIng committed:
v7.3

+ DietPi-Software | K3s: Do not overwrite existing config file on reinstall, especially since we do not (re)move /boot/dietpi-k3s.yaml but only copy it
+ DietPi-Software | K3s: Skip the obsolete separate G_CHECK_URL, since curl gives enough info when the download itself fails. G_CHECK_URL is still used in the script elsewhere, but will be gradually removed when touching code lines, probably in favour of a G_DOWNLOAD function to include some download/network specific troubleshoot/resolution options right into the error handler.
+ DietPi-Software | K3s: Pre-create /usr/local/bin, which might not always exist yet, but the installer expects it
+ DietPi-Software | K3s: On uninstall, check only for existing uninstall scripts, not if they are executable. If they exist but are not executable, then some serious mistake would have happened which is good to recognise better earlier than later, so it's good to run into an error instead of skipping the ordinary K3s uninstall silently.

MichaIng committed:
v7.3

+ DietPi-Software | K3s: Switch to software ID 193, since 192 is used by Snapcast Client already
+ DietPi-Software | K3s: Add apparmor and iptables to dependencies, otherwise K3s will run into a startup error
+ DietPi-Software | K3s: Do not enable and start the systemd service, as it will controlled by DietPi-Services instead
+ DietPi-Software | K3s: Uninstall K3s, if install fails and user chooses to ignore, as most likely files have been installed
+ DietPi-Software | K3s: Change systemd service restart behaviour so that it only restarts on successful exit (internal update?) but not on failure. Our position here is usually that a service failure means that the admin needs to fix an issue, instead of having a service constantly restarting, running into the same error again and again, spamming the logs and causing unnecessary system load, probably worsen file system corruption and similar. It needs to be observed if the K3s parent service intentionally terminates uncleanly in some circumstances and relies on being restarted by systemd then.
